xclim/indices/_threshold.py

```diff
@@ -1340,10 +1361,18 @@ def first_snowfall(
         First day of the year when the solid precipitation is superior to a threshold.
         If there is no such day, returns np.nan.
 
+    Notes
+    -----
+        The estimated mean snow density value of 312 kg m-3 is taken from :cite:t:`sturm_swe_2010`.
+
     References
     ----------
     :cite:cts:`cbcl_climate_2020`.
     """
+    unit = str2pint("mm/day")
+    unit_thresh = str2pint(thresh)
+    if unit_thresh._units == unit._units:
+        prsn = prsn_to_mm_per_day(prsn, snr=snr, const=const)
     thresh = convert_units_to(thresh, prsn, context="hydro")
```

Comment:
Similar problem here with `context="hydro"` and having `prsn` with units `mm day-1`. 

If we have `prsn [mm day-1]` and `thresh [kg m-2 day-1]`, then `thresh = convert_units_to(thresh, prsn, context="hydro")` will do a conversion that is not consistent with the choice of `const=312 kg m-3`. 

On the other hand, having both `prsn [mm day-1]` and `thresh [mm day-1]` means that we would use `prsn [mm day-1]` as an input in `prsn_to_mm_per_day`, so this is a concrete occurence of the problem outlined in my comment above.

Reply:
We could 
1) Restrict the use of the indicator `first_snowfall` to `prsn` with  [mass]/[area]/[time] units. (also, in this case,   `context="hydro" would not be necessary).
2) Allow `prsn` with  [length]/[time] units.  We should then figure out what we do when we have  `prsn [mm day-1]` and `thresh [kg m-2 day-1]`. We could convert  `prsn` with a similar function `prsn_to_kg_per_m2_per_day` (with some better  name)

Maybe taking `thresh` with units ` [kg m-2 day-1]`, so maybe it just complicates things for nothing. Or maybe it's really unusual to have `prsn` with `[mm day-1]` to start with, and sticking with option 1) would make sense. 

Anyways, I think we need to restrict use cases (1) or make sure every case is covered correctly (2).